### PR TITLE
[Authorization] Resend Email Verification

### PR DESCRIPTION
**Pull request by AnastasiiaVeselovska**
## Summary  
Implemented the **Resend Verification Email** feature and updated the `User` model to support pending email confirmation.

## Changes

- **Added `ResendEmailView`** (`users/views.py`) to allow resending verification emails without requiring authentication.
- **Updated `users/models.py`**:
  - Added `pending_email` field to temporarily store email changes before verification.
  - Added `confirm_pending_email()` method to finalize pending email updates after successful verification.
  - Removed `role` from `REQUIRED_FIELDS` to simplify user creation.
- Integrated email sending logic with support for `pending_email` override.
- Configured endpoint to always return a generic success message (prevents user enumeration).
- Ensured token generation is compatible with `EmailVerificationTokenGenerator`.
- Added support for optional `token` in the request payload.

## New Endpoint  
`POST /api/v1/auth/resend-email/`  
(See README for payload and usage details.)

## Other Notes

- Endpoint works without authentication.
- Throttling is disabled for this endpoint.
- Emails are sent in both HTML and plain-text formats.

image
image
image

---

> **This section should be added to `README.md`**

### Endpoint

```
POST /api/v1/auth/resend-email/
```

### Description

Resends a verification email to the user. The response is always generic to prevent user enumeration. You can optionally override the user's email.

### Payload (JSON)

| Field     | Type    | Required | Description                                            |
| --------- | ------- | -------- | ------------------------------------------------------ |
| `user_id` | integer | yes      | ID of the user to resend verification email.           |
| `email`   | string  | no       | Optional new email to update the user's pending email. |
| `token`   | string  | no       | Optional pre-generated verification token.             |

### Example Request

```json
{
    "user_id": 1,
    "email": "newemail@example.com"
}
```

### Example Response (HTTP 202 Accepted)

```json
{
    "detail": "If the account exists, a verification email has been sent."
}
```

### Notes

- Email content is sent in both HTML and plain text.
- `pending_email` is updated if a new email is provided.
- The actual verification link uses `uidb64` + `token` and will be included in the email.
- No authentication required.
- Throttled by default; can be configured via DRF throttling settings.

### CODE REVIEW

File: tests/users/test_resend_email.py
Comment (mehalyna) [inline tests/users/test_resend_email.py:1]: You are patching django.core.mail.send_mail and users.views.send_mail in different tests. For consistency and to avoid missing a patch, consider always patching the same location, and use the actual import path as used in the code under test.

Comment (mehalyna) [inline tests/users/test_resend_email.py:1]: Line 33, 47, 59, 75, 87, 105, 117:

The data dictionary includes a 'token' key with an empty string. If this is not required by the API, consider omitting it for clarity.
If the API expects only some of these fields, tailor each test case to send only what's required.

Comment (mehalyna) [inline tests/users/test_resend_email.py:1]: Line 41, 65, 126:

You are checking that user.pending_email is None or matches a value. Consider also asserting that user.email is still correct, especially in cases where the email should not change.

File: tests/users/test_resend_email.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+@pytest.mark.django_db
+@patch('django.core.mail.send_mail')
+def test_unknown_user_id_returns_202_no_exception(mock_send_mail, api_client):
+    url = reverse('resend-email')
+    data = {'user_id': 999999, 'token': ''}
+
+    response = api_client.post(url, data, format='json')
+
+    assert response.status_code == 202
+    mock_send_mail.assert_not_called()
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:52]: Consider adding an assertion to check the response data (e.g., that there is no error message or that the response body is empty as expected for 202).

File: tests/users/test_resend_email.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        role=user_role,
+        is_active=True
+    )
+    url = reverse('resend-email')
+    data = {'user_id': user.user_id, 'token': ''}
+
+    response = api_client.post(url, data, format='json')
+
+    assert response.status_code == 202
+    mock_send_mail.assert_not_called()
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:110]: Consider adding an assertion to check the response data (e.g., that there is no error message or that the response body is empty as expected for 202).

File: tests/users/test_resend_email.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+def test_throttling(mock_send_mail, api_client, user):
+    url = reverse('resend-email')
+    data = {'user_id': user.user_id, 'token': ''}
+
+    for _ in range(5):
+        response = api_client.post(url, data, format='json')
+        assert response.status_code == 202
+    
+    response = api_client.post(url, data, format='json')
+    assert response.status_code == 202
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:82]: You are making multiple requests and expecting 202 each time. If throttling is expected to eventually block or delay requests, consider asserting for a 429 status or whatever is appropriate for your API after the limit is exceeded.

File: tests/users/test_resend_email.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+@pytest.mark.django_db
+def test_bad_input_invalid_email(api_client, user):
+    url = reverse('resend-email')
+    data = {'user_id': user.user_id, 'email': 'not-an-email', 'token': ''}
+
+    response = api_client.post(url, data, format='json')
+
+    assert response.status_code == 400
+    assert 'email' in response.data
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:92]: You assert response.status_code == 400 and 'email' in response.data. Consider checking the specific error message, e.g. response.data['email'] contains the expected error string, to ensure correct validation.

File: tests/users/test_resend_email.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+@pytest.mark.django_db
+@patch('django.core.mail.send_mail')
+def test_already_verified_user_returns_202_no_email_sent(mock_send_mail, api_client, user_role):
+    user = User.objects.create(
+        email='active@example.com',
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:98]: When creating the already-verified user, you do not set email_verification_token or email_verification_sent_at. If your application logic relies on these fields, consider setting them to test the edge case more thoroughly.

File: tests/users/test_resend_email.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+@patch('users.views.send_mail')
+def test_email_override_saves_pending_and_sends_to_override(mock_send_mail, api_client, user):
+    url = reverse('resend-email')
+    new_email = 'newemail@example.com'
+    data = {'user_id': user.user_id, 'email': new_email, 'token': ''}
+
+    response = api_client.post(url, data, format='json')
+
+    print("Response status:", response.status_code)
+    print("Response data:", response.data)
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:122]: The print statements are helpful for debugging but should be removed or replaced with logging if this code is to be committed.

File: tests/users/test_resend_email.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+    user.refresh_from_db()
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:124]: You have nearly identical tests at lines 56 and 113. Consider combining or parametrizing these tests to reduce duplication.

File: users/tokens.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        """
+        base = f"{getattr(user, 'pk', '')}{timestamp}{getattr(user, 'is_active', False)}"
+
+        rotation_token = getattr(user, 'email_verification_token', None)
+        if rotation_token:
+            base += str(rotation_token)
+
+        pending_email = getattr(user, 'pending_email', None)
+        if pending_email:
+            base += str(pending_email)
```
Comment (mehalyna) [inline users/tokens.py:42]: The concatenation approach is simple, but consider using a separator (e.g., ":") to avoid possible collisions if values concatenate to the same string (e.g., "12" + "34" == "1" + "234").

File: users/tokens.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+        return base
+
+
+email_verification_token = EmailVerificationTokenGenerator()
```
Comment (mehalyna) [inline users/tokens.py:47]: Consider using UPPERCASE (EMAIL_VERIFICATION_TOKEN) for constants to distinguish from other variables.

File: users/tokens.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    Encode user_id to a URL-safe base64 string.
+
+    Args:
+        user_id (int): User primary key.
+
+    Returns:
+        str: Base64 encoded user ID.
+    """
+    return urlsafe_base64_encode(force_bytes(user_id))
```
Comment (mehalyna) [inline users/tokens.py:60]: You return urlsafe_base64_encode(force_bytes(user_id)). If user_id is not an int, this may fail or encode incorrectly. Consider type-checking or converting to str first.

File: users/tokens.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    Decode a base64 string back to user_id.
+
+    Args:
+        uidb64 (str): Base64 encoded user ID.
+
+    Returns:
+        int or None: Decoded user ID, or None if decoding fails.
+    """
+    try:
+        return int(force_str(urlsafe_base64_decode(uidb64)))
```
Comment (mehalyna) [inline users/tokens.py:74]: The decode logic assumes the result is always an integer. If the underlying user_id is not an int, this could raise an exception or produce an incorrect result. Consider validating the decoded value.

File: users/tokens.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    except (TypeError, ValueError, OverflowError):
+        return None
```
Comment (mehalyna) [inline users/tokens.py:76]: You catch TypeError, ValueError, and OverflowError. Consider logging or raising a custom exception for easier debugging, instead of silently returning None.

File: users/tokens.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    Generate an email verification token for the user.
+
+    Args:
+        user (User): User instance.
+
+    Returns:
+        str: Token string.
+    """
+    return email_verification_token.make_token(user)
```
Comment (mehalyna) [inline users/tokens.py:89]: No input validation: if user is None or malformed, this will raise. Consider asserting or raising a helpful error early.

File: users/tokens.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    Validate the token for the user.
+
+    Args:
+        user (User): User instance.
+        token (str): Token string.
+
+    Returns:
+        bool: True if valid, False otherwise.
+    """
+    return email_verification_token.check_token(user, token)
```
Comment (mehalyna) [inline users/tokens.py:103]: You delegate to email_verification_token.check_token. Confirm that this handles all relevant edge cases (inactive users, empty tokens, etc.), or document any limitations here.

File: users/views.py [hunk 15]
```diff
@@ -26,6 +26,12 @@
 from rest_framework.permissions import AllowAny
 from rest_framework.throttling import AnonRateThrottle
 from rest_framework_simplejwt.views import TokenObtainPairView
+from rest_framework.permissions import AllowAny
+from rest_framework.throttling import ScopedRateThrottle
+from django.utils.http import urlsafe_base64_encode
+from django.utils.encoding import force_bytes
+from .serializers import ResendEmailSerializer
+from .tokens import email_verification_token
```
Comment (mehalyna) [inline users/views.py:34]: You're importing AllowAny twice (lines 26 and 29); remove the duplicate import for clarity.
Imports are not grouped according to convention (standard library, third-party, local); consider organizing them for readability.

File: users/views.py [hunk 16]
```diff
@@ -142,6 +148,12 @@ def get(self, request, user_id, token):
                 email_verification_token=token
             )
             
+            if user.is_active:
+                return Response(
+                    {'status': 'success', 'message': 'Email is already verified.'},
+                    status=status.HTTP_200_OK
+                )
+
```
Comment (mehalyna) [inline users/views.py:156]: Returning a 200 when the user is already verified is good, but you might want to clarify in the response if further action is possible (e.g., "No email was sent because the address is already verified").

File: users/views.py [hunk 17]
```diff
@@ -155,6 +167,10 @@ def get(self, request, user_id, token):
                     status=status.HTTP_400_BAD_REQUEST
                 )
             
+            if user.pending_email:
+                user.email = user.pending_email
+                user.pending_email = None
+
```
Comment (mehalyna) [inline users/views.py:173]: You're updating user.email and clearing pending_email directly in the view.
Move this logic to a dedicated User model method (e.g., confirm_pending_email) to improve maintainability and reuse, and ensure business rules are enforced consistently.
After updating emails, consider logging the change or triggering audit signals if required by your application.

File: users/views.py [hunk 18]
```diff
@@ -179,6 +195,108 @@ def get(self, request, user_id, token):
...
+        """
+        serializer = ResendEmailSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        user_id = serializer.validated_data["user_id"]
+        email = serializer.validated_data.get("email")
+        token = serializer.validated_data.get("token")
+
+        try:
+            user = User.objects.get(user_id=user_id)
```
Comment (mehalyna) [inline users/views.py:241]: Instead of returning immediately if user.is_active, consider still updating the pending_email if a new one is provided (depending on business logic).

File: users/views.py [hunk 19]
```diff
@@ -179,6 +195,108 @@ def get(self, request, user_id, token):
...
+
+        except User.DoesNotExist:
+            return Response(
+                {"detail": "If the account exists, a verification email has been sent."},
+                status=status.HTTP_202_ACCEPTED,
+            )
+
+        if email:
+            user.pending_email = email
+            user.save(update_fields=["pending_email"])
```
Comment (mehalyna) [inline users/views.py:256]: When updating user.pending_email, consider normalizing the email (strip and lower) before saving, to prevent duplicate logical email addresses.
Use a try/except around user.save() to handle possible database errors (e.g., unique constraint violations).

File: users/views.py [hunk 20]
```diff
@@ -179,6 +195,108 @@ def get(self, request, user_id, token):
...
+        else:
+            email = user.pending_email or user.email
```
Comment (mehalyna) [inline users/views.py:258]: mail = user.pending_email or user.email: if both are None (corrupted data), handle gracefully with an error or warning.

File: users/views.py [hunk 21]
```diff
@@ -179,6 +195,108 @@ def get(self, request, user_id, token):
...
+        verification_relative_url = reverse(
+            'verify-email', kwargs={'user_id': user.user_id, 'token': token}
+        )
+        verify_url = f"{settings.FRONTEND_URL}{verification_relative_url}?email={email}"
+
+        context = {
+            'user': user,
+            'verification_url': verify_url,
+            'user_display_name': user.first_name or user.email,
+            'support_text': "If you didn't request this, please ignore this email.",
```
Comment (mehalyna) [inline users/views.py:272]: support_text could be moved to a constant or settings file if reused elsewhere.

File: users/views.py [hunk 22]
```diff
@@ -179,6 +195,108 @@ def get(self, request, user_id, token):
...
+        }
+
+        subject = "Confirm your email"
+        html_message = render_to_string('email/activation.html', context)
```
Comment (mehalyna) [inline users/views.py:276]: Template path 'email/activation.html' is hardcoded - consider moving to a setting or constant for flexibility.

File: users/views.py [hunk 23]
```diff
@@ -179,6 +195,108 @@ def get(self, request, user_id, token):
...
+
+        try:
+            send_mail(
+                subject=subject,
+                message=plain_message,
+                from_email=settings.DEFAULT_FROM_EMAIL,
+                recipient_list=[email],
+                html_message=html_message,
+                fail_silently=False,
+            )
```
Comment (mehalyna) [inline users/views.py:291]: If sending fails, the exception is logged but not surfaced to the user. Consider incrementing a metric or alerting for operational visibility.

File: users/views.py [hunk 24]
```diff
@@ -179,6 +195,108 @@ def get(self, request, user_id, token):
...
+        except Exception as e:
+            logger.error(f"Failed to send verification email to {email}: {str(e)}")
+
+        return Response(
+            {"detail": "If the account exists, a verification email has been sent."},
+            status=status.HTTP_202_ACCEPTED,
+        )
+
```
Comment (mehalyna) [inline users/views.py:299]: Always returning 202 regardless of errors may mask repeated internal failures; consider returning a 500 in obvious misconfiguration cases.

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: tests/users/test_resend_email.py (hunk 1 above)
Comment (MaksymZymyn1989) [reply tests/users/test_resend_email.py:52]: Consider not using pytest.
The command "python manage.py test" doesn't see these tests.
Now we don't have these tests in CI/Cd pipeline.

File: tests/users/test_resend_email.py
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py:1]: I tried to address the comments, but I rewrote the tests using `Django APITestCase`.

File: tests/users/test_resend_email.py (hunk 1 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: I tried to address the comments, but I rewrote the tests using `Django APITestCase`.

File: tests/users/test_resend_email.py (hunk 3 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: I tried to address the comments, but I rewrote the tests using `Django APITestCase`.

File: tests/users/test_resend_email.py (hunk 4 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: I tried to address the comments, but I rewrote the tests using `Django APITestCase`.

File: tests/users/test_resend_email.py (hunk 5 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: I tried to address the comments, but I rewrote the tests using `Django APITestCase`.

File: tests/users/test_resend_email.py (hunk 6 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: I tried to address the comments, but I rewrote the tests using `Django APITestCase`.

File: tests/users/test_resend_email.py (hunk 7 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: I tried to address the comments, but I rewrote the tests using `Django APITestCase`.

File: users/models.py [hunk 25]
```diff
@@ -367,6 +369,51 @@ def get_all(cls):
...
+            raise ValidationError("No pending email to confirm.")
+
+        if User.objects.filter(email__iexact=self.pending_email).exclude(pk=self.pk).exists():
+            raise ValidationError("This email is already in use by another user.")
+
+        self.email = self.pending_email
+        self.pending_email = None
+        self.email_verification_sent_at = None
+        self.save(update_fields=['email', 'pending_email', 'email_verification_sent_at'])
+
```
[the remark this answers is not in the document]
Comment (AnastasiiaVeselovska) [reply users/models.py]: Thank you. The `verified_at` field was added and the `confirm_pending_email` method updated.

File: users/tokens.py [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+        Construct a unique hash value for token generation.
+
+        Args:
+            user (User): The user instance.
+            timestamp (int): Token generation timestamp.
+
+        Returns:
+            str: String uniquely identifying the token for the user.
+        """
+        base = f"{getattr(user, 'pk', '')}{timestamp}{getattr(user, 'is_active', False)}"
```
[the remark this answers is not in the document]
Comment (AnastasiiaVeselovska) [reply users/tokens.py]: Thank you. Added explicit `str` cast for `pk` to ensure consistent token generation.

File: users/tokens.py (hunk 9 above)
Comment (AnastasiiaVeselovska) [reply users/tokens.py]: Thank you. I've updated the variable to a constant in uppercase (`EMAIL_VERIFICATION_TOKEN`) to clearly distinguish it from other variables.

File: users/tokens.py (hunk 10 above)
Comment (AnastasiiaVeselovska) [reply users/tokens.py]: Thank you. Casted `user_id` to string before encoding to ensure `urlsafe_base64_encode` works correctly with any type.

File: users/tokens.py (hunk 11 above)
Comment (AnastasiiaVeselovska) [reply users/tokens.py]: Thank you. Make `decode_uidb64` safely handle non-integer user IDs.

File: users/tokens.py (hunk 13 above)
Comment (AnastasiiaVeselovska) [reply users/tokens.py]: Thank you. Added explicit checks for `None` users in `make_token and check_token` to fail early with a clear error.

File: users/tokens.py (hunk 14 above)
Comment (AnastasiiaVeselovska) [reply users/tokens.py]: Thank you. Add explicit errors for missing user or token in `check_token`.

File: users/views.py (hunk 17 above)
Comment (AnastasiiaVeselovska) [reply users/views.py]: Thank you. Moved email update to `User.confirm_pending_email()` so the view no longer modifies email or `pending_email` directly. Logging is now handled in the model.

File: users/views.py (hunk 19 above)
Comment (AnastasiiaVeselovska) [reply users/views.py]: Thank you. Normalize `pending_email` before saving and handle possible `IntegrityError` when updating.

File: users/views.py (hunk 20 above)
Comment (AnastasiiaVeselovska) [reply users/views.py]: Thank you. Added a check for missing emails. If both `pending_email` and email are `None`, a warning is logged and a 400 response is returned to prevent attempting to send an email to `None`.

File: users/views.py (hunk 21 above)
Comment (AnastasiiaVeselovska) [reply users/views.py]: Thank you. Replaced inline string with `SUPPORT_TEXT` constant from `constants.py` for reuse and easier maintenance.

File: users/views.py (hunk 22 above)
Comment (AnastasiiaVeselovska) [reply users/views.py]: Thank you. The template path was hardcoded, but it’s now moved to a constant to improve maintainability and flexibility.

File: users/views.py (hunk 23 above)
Comment (AnastasiiaVeselovska) [reply users/views.py]: Thank you. Updated to log as critical with full traceback for better visibility. Metrics or alerting can be added later for operational monitoring.

File: users/views.py (hunk 20 above)
Comment (AnastasiiaVeselovska) [reply users/views.py]: Thank you. Added a check for missing emails. If both `pending_email` and email are `None`, a warning is logged and a 400 response is returned to prevent attempting to send an email to `None`.

File: tests/users/test_resend_email.py [hunk 27]
```diff
@@ -0,0 +1,151 @@
+import logging
+import uuid
+from django.urls import reverse
+from django.utils import timezone
+from rest_framework.test import APITestCase
+from unittest.mock import patch
+from users.models import User, UserRole
+
+logger = logging.getLogger(__name__)
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:9]: The logger is defined here, but this is a test file. Excessive logging in tests can produce noisy output. Consider using logging only for debugging or critical failures, and possibly remove info-level logs.

File: tests/users/test_resend_email.py [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+logging.basicConfig(level=logging.INFO)
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:10]: Configuring logging at the test file level may interfere with Django or pytest logging. If logging is necessary, configure it in a test runner or a dedicated conftest.py instead.

File: tests/users/test_resend_email.py [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+        self.user = User.objects.create(
+            email=f"user_{uuid.uuid4().hex[:10]}@example.com",
+            first_name='Test',
+            last_name='User',
+            role=self.user_role,
+            is_active=False,
+            email_verification_token='oldtoken',
+            email_verification_sent_at=timezone.now()
+        )
+        logger.info("Created test user: %s", self.user.email)
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:29]: Remove or reduce logging unless debugging. Test output should generally remain clean.

File: tests/users/test_resend_email.py [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+
+    def perform_resend_email_test(self, user_id, expected_status=202, email=None, user_obj=None):
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:31]: Consider using more descriptive parameter names, e.g., `target_user_id` instead of `user_id`, for clarity.

File: tests/users/test_resend_email.py [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+    def test_resend_email_scenarios(self, mock_make_token, mock_send_mail):
+        scenarios = [
+            ("happy_path", None, None, True),
+            ("email_override", "newemail@example.com", "newemail@example.com", True),
+            ("unknown_user", None, None, False),
+        ]
+
+        for i, (scenario, email, expected_pending_email, send_mail_expected) in enumerate(scenarios):
+            if scenario == "unknown_user":
+                user_id = 999999
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:61]: Instead of a magic number, consider using None or a constant with a descriptive name, e.g., `NON_EXISTENT_USER_ID`.

File: tests/users/test_resend_email.py [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+                    self.assertEqual(test_user.pending_email, expected_pending_email)
+                else:
+                    self.assertIsNone(test_user.pending_email)
+                self.assertTrue(test_user.email.endswith('@example.com'))
+
+            if send_mail_expected:
+                mock_send_mail.assert_called()
+                if email:
+                    _, kwargs = mock_send_mail.call_args
+                    self.assertIn(email, kwargs['recipient_list'])
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:91]: If `recipient_list` is a list, consider using assertIn for each recipient, and add an explicit check for `recipient_list` existence.

File: tests/users/test_resend_email.py [hunk 33]
```diff
@@ -0,0 +1,151 @@
...
+            else:
+                mock_send_mail.assert_not_called()
+
+            mock_send_mail.reset_mock()
+
+    @patch('users.views.ResendEmailView.throttle_classes', [])
+    def test_bad_input_invalid_email(self):
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:98]: Use parameterized tests for input validation to easily test multiple invalid email formats.

File: tests/users/test_resend_email.py [hunk 34]
```diff
@@ -0,0 +1,151 @@
...
+            self.assertIn(expected_error, str(errors))
+        self.assertTrue(self.user.email.startswith('user_'))
+        logger.info("Bad input test: invalid email correctly rejected")
+
+    @patch('users.views.send_mail')
+    @patch('users.views.EMAIL_VERIFICATION_TOKEN.make_token', return_value='newtoken')
+    def test_throttling_with_limit(self, mock_make_token, mock_send_mail):
+        url = reverse('resend-email')
+        data = {'user_id': self.user.user_id}
+        allowed_requests = 5
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:121]: This test assumes a throttling limit of 5. If this value is configurable, fetch from settings or document the assumption.

File: tests/users/test_resend_email.py [hunk 35]
```diff
@@ -0,0 +1,151 @@
...
+
+        for i in range(allowed_requests):
+            response = self.client.post(url, data, format='json')
+            self.assertEqual(response.status_code, 202)
+            self.user.refresh_from_db()
+            self.assertTrue(self.user.email.startswith('user_'))
+            logger.info("Throttling test: request %d successful", i + 1)
```
Comment (mehalyna) [inline tests/users/test_resend_email.py:128]: Add a subTest context for each iteration for better failure reporting.

File: users/models.py [hunk 36]
```diff
@@ -369,6 +377,54 @@ def get_all(cls):
         """
         return cls.active_users.all()
     
+    def confirm_pending_email(self):
```
Comment (mehalyna) [inline users/models.py:380]: Add @transaction.atomic decorator for clarity, rather than using with transaction.atomic() inside the method.

File: users/models.py [hunk 37]
```diff
@@ -369,6 +377,54 @@ def get_all(cls):
...
+            raise ValidationError({"pending_email": ["This email is already in use by another user."]}, code="email_taken")
+        
+        with transaction.atomic():
+            self.email = normalized_email
+            self.pending_email = None
+            self.email_verification_sent_at = None
+            self.verified_at = None
+            self.save(update_fields=['email', 'pending_email', 'email_verification_sent_at', 'verified_at'])
+        
+        logger.info(f"User {self.user_id} confirmed pending email. New email: {self.email}")
```
Comment (mehalyna) [inline users/models.py:407]: Log email changes at 'info' level only if you have user consent; consider using 'warning' or 'audit' logger for sensitive changes.

File: users/tokens.py [hunk 38]
```diff
@@ -0,0 +1,118 @@
...
+def decode_uidb64(uidb64: str) -> Optional[int]:
+    """
+    Decode a base64 string back to user_id.
+
+    Args:
+        uidb64 (str): Base64 encoded user ID.
+
+    Returns:
+        Optional[Union[int, str]]: Decoded user ID as int if possible, else str. None if decoding fails.
+    """
```
Comment (mehalyna) [inline users/tokens.py:77]: The return type and docstring are inconsistent (docstring mentions Optional[Union[int, str]], but annotation says Optional[int]). Update either the annotation or the docstring for clarity.

File: users/tokens.py [hunk 39]
```diff
@@ -0,0 +1,118 @@
...
+
+    Args:
+        user (User): User instance.
+
+    Returns:
+        str: Token string.
+    """
+    if user is None:
+        raise ValueError("Cannot generate token: user must not be None.")
+    return EMAIL_VERIFICATION_TOKEN.make_token(user)
```
Comment (mehalyna) [inline users/tokens.py:100]: Consider raising a ValueError if the user is inactive or has some other invalid state, if that fits your business logic.

File: users/views.py [hunk 40]
```diff
@@ -187,6 +224,137 @@ def get(self, request, user_id, token):
...
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        if not token:
+            token = EMAIL_VERIFICATION_TOKEN.make_token(user)
+
+        verification_relative_url = reverse(
+            'verify-email', kwargs={'user_id': user.user_id, 'token': token}
+        )
+        verify_url = f"{settings.FRONTEND_URL}{verification_relative_url}?email={email_to_send}"
```
Comment (mehalyna) [inline users/views.py:308]: Including the email in the query string may expose sensitive info. Consider if this is necessary.

**Review by mehalyna — APPROVED** [review]

File: tests/users/test_resend_email.py (hunk 30 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: Thank you. Updated the parameter names to be more descriptive (`target_user_id` instead of `user_id`) for clarity.

File: tests/users/test_resend_email.py (hunk 31 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: Thank you. I replaced the magic number with the constant `NON_EXISTENT_USER_ID` for better readability.

File: tests/users/test_resend_email.py (hunk 32 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: Thank you. I’ve refactored the tests to improve user creation and added explicit checks for `recipient_list` and `pending_email` to make the tests clearer and more robust.

File: tests/users/test_resend_email.py (hunk 33 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: Thank you. Converted single invalid email test into a parameterized test using `subTest` to cover multiple formats.

File: tests/users/test_resend_email.py (hunk 34 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: Thank you. Updated `ALLOWED_THROTTLE_REQUESTS` to be fetched from Django settings (`API_THROTTLE_LIMIT`) with a default of 5, so the test adapts automatically if the throttling limit changes.

File: tests/users/test_resend_email.py (hunk 35 above)
Comment (AnastasiiaVeselovska) [reply tests/users/test_resend_email.py]: Thank you. I’ve added `subTest` contexts for each iteration in `test_resend_email_scenarios` and `test_throttling_with_limit` to improve failure reporting.

File: users/models.py (hunk 37 above)
Comment (AnastasiiaVeselovska) [reply users/models.py]: Thank you. Corrected email change logs to use `info` and `warning` appropriately.

File: users/tokens.py (hunk 38 above)
Comment (AnastasiiaVeselovska) [reply users/tokens.py]: Thank you. Updated the function annotation to `Optional[Union[int, str]]` to match the docstring.

File: users/views.py (hunk 40 above)
Comment (AnastasiiaVeselovska) [reply users/views.py]: Thank you. Removed email from the query string to avoid exposing sensitive information.
